### PR TITLE
[Pull Request] Add implementation of Gaussian DP

### PR DESCRIPTION
**Pull request by woodyx218**
Add privacy accountants for Gaussian Differential Privacy (GDP) from https://arxiv.org/abs/1911.11607 as GDprivacy_accountants.py in tensorflow_privacy/privacy/analysis. The accountants calculate GDP of optimizers with uniform and Poisson subsampling.

Add two tutorials in the folder tutorials. They correspond to experiments from the same paper: Movielens 1M and IMDB movie reviews until certain GDP budget runs out. The experiments are written in a very similar way to tutorial/mnist_dpsgd_tutorial.py except using different neural networks, a different optimizer and computing GDP instead of epsilon-DP.

### CODE REVIEW

**Comment by woodyx218** [comment]
I have used pylint to reformat my code as well as modify the code according to your suggestion. Shall I push again (pull request again)?

File: tensorflow_privacy/privacy/analysis/gdp_accountant.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+# Target delta:delta
+
+def compute_mu_uniform(epoch, noise_multi, N, batch_size):
+    '''Compute mu from uniform subsampling'''
+    T = epoch*N/batch_size
+    c = batch_size*np.sqrt(T)/N
+    return np.sqrt(2)*c*np.sqrt(np.exp(noise_multi**(-2))*\
+                   norm.cdf(1.5/noise_multi)+3*norm.cdf(-0.5/noise_multi)-2)
+
+def compute_mu_Poisson(epoch, noise_multi, N, batch_size):
```
Comment (galenmandrew) [inline tensorflow_privacy/privacy/analysis/gdp_accountant.py]: lower case poisson

Comment (galenmandrew) [inline tensorflow_privacy/privacy/analysis/gdp_accountant.py]: lower  case poisson

File: tensorflow_privacy/privacy/analysis/gdp_accountant.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+from scipy import optimize
+
+# Total number of examples:N
+# batch size:batch_size
+# Noise multiplier for DP-SGD/DP-Adam:noise_multiplier
+# current epoch:epoch
+# Target delta:delta
+
+def compute_mu_uniform(epoch, noise_multi, N, batch_size):
+    '''Compute mu from uniform subsampling'''
```
Comment (galenmandrew) [inline tensorflow_privacy/privacy/analysis/gdp_accountant.py]: add period

File: tensorflow_privacy/privacy/analysis/gdp_accountant.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    return np.sqrt(2)*c*np.sqrt(np.exp(noise_multi**(-2))*\
+                   norm.cdf(1.5/noise_multi)+3*norm.cdf(-0.5/noise_multi)-2)
+
+def compute_mu_Poisson(epoch, noise_multi, N, batch_size):
+    '''Compute mu from Poisson subsampling'''
+    T = epoch*N/batch_size
+    return np.sqrt(np.exp(noise_multi**(-2))-1)*np.sqrt(T)*batch_size/N
+
+def delta_eps_mu(eps, mu):
+    '''Dual between mu-GDP and (epsilon, delta)-DP'''
```
Comment (galenmandrew) [inline tensorflow_privacy/privacy/analysis/gdp_accountant.py]: All single-line descriptions should be like "Computes dual between ..." It should say what the function *does* and end with a period.

File: tensorflow_privacy/privacy/analysis/gdp_accountant.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    return norm.cdf(-eps/mu+mu/2)-np.exp(eps)*norm.cdf(-eps/mu-mu/2)
+
+def eps_from_mu(mu, delta):
+    '''inverse Dual'''
```
Comment (galenmandrew) [inline tensorflow_privacy/privacy/analysis/gdp_accountant.py]: Computes inverse dual from mu and delta.

File: tensorflow_privacy/privacy/analysis/gdp_accountant.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    def f(x):
+        '''reversely solving dual'''
+        return delta_eps_mu(x, mu) - delta
+    return optimize.root_scalar(f, bracket=[0, 500], method='brentq').root
+
+def compute_eps_uniform(epoch, noise_multi, N, batch_size, delta):
+    '''inverse Dual of uniform subsampling'''
```
Comment (galenmandrew) [inline tensorflow_privacy/privacy/analysis/gdp_accountant.py]: Computes inverse dual for uniform subsampling.

File: tutorials/adult_tutorial.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+                   'Ratio of the standard deviation to the clipping norm')
+flags.DEFINE_float('l2_norm_clip', 1, 'Clipping norm')
+flags.DEFINE_integer('epochs', 20, 'Number of epochs')
+flags.DEFINE_integer('max_mu', 2, 'GDP upper limit')
+flags.DEFINE_string('model_dir', None, 'Model directory')
+
+microbatches = 256
+
+def nn_model_fn(features, labels, mode):
+    ''' Define CNN architecture using tf.keras.layers.'''
```
Comment (galenmandrew) [inline tutorials/adult_tutorial.py]: nit: extra space char

File: tutorials/adult_tutorial.py [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+from absl import app
+from absl import flags
+
+import numpy as np
+import tensorflow as tf
+import pandas as pd
+from sklearn.model_selection import KFold
+
+# from tensorflow_privacy.privacy.analysis.rdp_accountant import compute_rdp
+# from tensorflow_privacy.privacy.analysis.rdp_accountant import get_privacy_spent
```
Comment (galenmandrew) [inline tutorials/adult_tutorial.py]: Remove commented out code.

File: tutorials/adult_tutorial.py [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+
+    # Create tf.Estimator input functions for the training and test data.
+    eval_input_fn = tf.compat.v1.estimator.inputs.numpy_input_fn(
+        x={'x': test_data},
+        y=test_labels,
+        num_epochs=1,
+        shuffle=False)
+
+    # Training loop.
+    steps_per_epoch = 29305 // 256
```
Comment (galenmandrew) [inline tutorials/adult_tutorial.py]: make this a named constant like "num_examples"

File: tutorials/adult_tutorial.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    test_accuracy_list = []
+    for epoch in range(1, FLAGS.epochs + 1):
+        for step in range(steps_per_epoch):
+            whether = np.random.random_sample(29305) > (1-256/29305)
```
Comment (galenmandrew) [inline tutorials/adult_tutorial.py]: sample < microbatch_size would be more intuitive

File: tutorials/adult_tutorial.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+            subsampling = [i for i in np.arange(29305) if whether[i]]
+            global microbatches
+            microbatches = len(subsampling)
```
Comment (galenmandrew) [inline tutorials/adult_tutorial.py]: this is confusing. Why do you need to set the global microbatches? Probably there is a safer way to do this.

File: tutorials/movielens_tutorial.py [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+                                                           dtype=tf.float32),
+                                             axes=1))
+        }
+        return tf.estimator.EstimatorSpec(mode=mode,
+                                          loss=scalar_loss,
+                                          eval_metric_ops=eval_metric_ops)
+    return None
+
+
+def load_adult():
```
Comment (galenmandrew) [inline tutorials/movielens_tutorial.py]: should this be "load_movielens"?

**Comment by woodyx218** [comment]
Thank you @galenmandrew ! This is a great learning experience. I have changed the unnamed constants and clear the misspelling. The codes are tested again on my side and should be free of problems that you mentioned. The reason we have a global microbatch is because in each step, we are using a possibly different sized batch due to Poisson subsampling.

**Comment by galenmandrew** [comment]
Thanks @woodyx218 . This looks good to me. I just want to let @npapernot have a chance to take a look before approving.

**Comment by woodyx218** [comment]
Hi @npapernot and @galenmandrew, I understand the approval takes time and hope everything works fine. If you need any fixes from my side, I am happy to work on it right away.

**Comment by woodyx218** [comment]
Hi @galenmandrew , may I ask what's the status now and whether I need to work on next steps?

**Comment by npapernot** [comment]
I think it would be best to have less tutorials. Having one tutorial that demonstrates the new functionality is ok, but could you move the other tutorials (if they just change the dataset) to the `research/` folder in this repository?

**Comment by woodyx218** [comment]
Thank you for the feedback. I have moved two tutorials 'adult_tutorial.py' and 'imdb_tutorial.py' to research/GDP_2019 folder. Please let me know whether I need to modify the README.md in the research folder or the tutorial folder.

File: tensorflow_privacy/privacy/analysis/gdp_accountant.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+Theorem (CLT) to estimate privacy budget of an iterated subsampled
+Gaussian Mechanism (by either uniform or Poisson subsampling).
+"""
+
+
+import numpy as np
+from scipy.stats import norm
+from scipy import optimize
+
+# Total number of examples:N
```
Comment (npapernot) [inline tensorflow_privacy/privacy/analysis/gdp_accountant.py]: Move these doc strings within each function below

File: tensorflow_privacy/privacy/analysis/gdp_accountant.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+# batch size:batch_size
+# Noise multiplier for DP-SGD/DP-Adam:noise_multiplier
+# current epoch:epoch
+# Target delta:delta
+
+def compute_mu_uniform(epoch, noise_multi, N, batch_size):
```
Comment (npapernot) [inline tensorflow_privacy/privacy/analysis/gdp_accountant.py]: Add doc string for each function with parameters

File: tutorials/movielens_tutorial.py [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+from absl import app
+from absl import flags
+
+import numpy as np
+import tensorflow as tf
+import pandas as pd
+from scipy.stats import rankdata
+from sklearn.model_selection import train_test_split
+
+from tensorflow_privacy.privacy.optimizers import dp_optimizer
```
Comment (npapernot) [inline tutorials/movielens_tutorial.py]: Why empty line here?

File: tutorials/movielens_tutorial.py [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+
+from tensorflow_privacy.privacy.analysis.gdp_accountant import *
```
Comment (npapernot) [inline tutorials/movielens_tutorial.py]: Can you add explicit import names?

**Comment by woodyx218** [comment]
@npapernot I have modified my codes according to your advices! Thank you again for quick feedback. Looking forward to hearing from you again.

**Comment by npapernot** [comment]
Can you look at the doc string styles in the rest of the library and follow the same style?

**Comment by woodyx218** [comment]
Hi I did check the rest of the library and gdp_accountant.py is the only place that needs modification. I also corrected other tutorials.

**Comment by woodyx218** [comment]
Hi guys. Thanks again for approving this. May I check when will the GDP part be merged, roughly speaking? We are trying to advocate our method so it would be nice to share the Tensorflow code with people!

**Comment by woodyx218** [comment]
Hi guys, may I get a rough date about merging our pull request? It has been a while since the request was approved and we hope to update our friends in differential privacy about this. Best wishes!

**Comment by galenmandrew** [comment]
woodyx218-- the change is committed. Thank you for your contribution and please accept our apologies for the latency!

One thing: although you are credited in the comment for the change, due to a problem in our commit process the change lists the author as being only "tensorflower-gardener". Is this important to you? If so it may be possible to roll back and recommit listing you as author.

**Comment by woodyx218** [comment]
Hi if it's not too much trouble, I would like to be credited as the author. Thank you in advance!

**Comment by schien1729** [comment]
Okay, I'll see what we can do.

**Comment by schien1729** [comment]
All right, I think this is fixed.

**Pull request merged**